### PR TITLE
fix https://github.com/elves/elvish/issues/1220

### PR DESCRIPTION
**Pull request by mrlzh**
```
if j>len(name){
	break
}
```
Check if j is greater than name's length

### CODE REVIEW

File: pkg/glob/glob.go [hunk 1]
```diff
@@ -228,6 +228,9 @@ segs:
 				if !startingStar.Match(r) {
 					break
 				}
+				if j > len(name) {
```
Comment (krader1961) [inline pkg/glob/glob.go:231]: Either this test should be moved above the prior `if !startingStar...` block or the `j := i + len(string(r))` statement should be moved to immediately prior to this line. It is unclear if the calculation of `j` should affect whether or not `startingStar.Match(r)` is called. I don't doubt this change fixes the panic but in its current form it introduces more ambiguity about the logic.

**Comment by xiaq** [comment]
This PR patches up the panic but it doesn't address the underlying correctness issue.

The underlying issue is that the code here assumes that

```go
for i, r := range s { ... }
```

Is always a lossless way to split a string, and all the `r`s, when concatenated together, will form `s`. This is only true if `s` is a valid UTF-8 string; if there is anything that can't be decoded as UTF-8, the iteration will advance by one byte with `r = 0xFFFD` (whose length is not 1 when converted to a string, which explains the panic). OS paths are not guaranteed to be valid UTF-8 strings, and the reproduction in #1220 is essentially constructing a non-UTF-8 string.

The fix should be changing the `glob` package to treat strings as byte sequences instead of rune sequences.

**Comment by krader1961** [comment]
@mrlzh, See also the description of the [two ways to iterate over a string](https://yourbasic.org/golang/for-loop-range-array-slice-map-channel/#string-iteration-runes-or-bytes). @xiaq is correct that Elvish cannot assume path names are valid UTF-8 strings. But since the glob/wildcard characters are single byte ASCII characters that doesn't matter. It is sufficient to process the path name one byte at at time. If the byte is a wildcard (e.g., `*`) it is treated specially; otherwise, the byte is a literal that has to match the corresponding byte in any expansion.

**Comment by xiaq** [comment]
The bug is now fixed in e061b657cbd1621f420d5f85d6626d4bea7864bb, closing.

**Pull request closed without merging**